aidial_adapter_openai/utils/streaming.py

```diff
@@ -61,84 +61,112 @@ async def generate_stream(
+        chunk = chunk or noop_chunk
+        chunk["statistics"] = {"discarded_messages": indices}
+        return chunk
+
+    n_chunks = 0
+    last_chunk = None
+    buffer_chunk = None
+
+    completions: dict[int, str] = {}
+    found_finish_reason = False
+    found_usage = False
+    error = None
 
     try:
         async for chunk in stream:
-            if len(chunk["choices"]) > 0:
-                if temp_chunk is not None:
-                    chunk = merge(temp_chunk, chunk)
-                    temp_chunk = None
+            n_chunks += 1
+
+            if buffer_chunk is not None:
+                chunk = merge_chunks(buffer_chunk, chunk)
+                buffer_chunk = None
+
+            choices = chunk.get("choices") or []
 
-                choice = chunk["choices"][0]
-                total_content += (choice.get("delta") or {}).get(
-                    "content"
-                ) or ""
+            for choice in choices:
+                index = choice["index"]
+                content = (choice.get("delta") or {}).get("content") or ""
 
-                if choice["finish_reason"] is not None:
-                    stream_finished = True
-                    finalize_finish_chunk(chunk)
+                completions[index] = completions.get(index, "") + content
+                found_finish_reason |= bool(choice.get("finish_reason"))
 
-                yield chunk
+            found_usage |= bool(chunk.get("usage"))
+
+            # Azure OpenAI returns an empty list of choices as a first chunk
+            # when content filtering is enabled for a corresponding deployment.
+            # The safety rating of the request is reported in this first chunk.
+            # Here we withhold such a chunk and merge it later with a follow-up chunk.
+            if len(choices) == 0 and eliminate_empty_choices:
+                buffer_chunk = chunk
             else:
-                if fix_streaming_issues_in_new_api_versions:
-                    temp_chunk = chunk
-                else:
-                    yield chunk
+                if last_chunk is not None:
+                    yield last_chunk
+                last_chunk = chunk
 
-            last_chunk = chunk
     except APIError as e:
         status_code = e.status_code if isinstance(e, APIStatusError) else 500
-        yield DialException(
+        error = DialException(
             status_code=status_code,
             message=e.message,
             type=e.type,
             param=e.param,
             code=e.code,
         ).json_error()
-        return
 
-    if not stream_finished:
-        if last_chunk is None:
+    if last_chunk is not None and buffer_chunk is not None:
+        last_chunk = merge_chunks(buffer_chunk, last_chunk)
+
+    if discarded_messages is not None:
+        last_chunk = set_discarded_messages(last_chunk, discarded_messages)
+
+    if not found_usage and (not error or completions):
```

Comment:
Here, if user received no errors, and no chunks, we would not return to him usage for prompt tokens, why?

If we'll leave just
```python
if not found_usage:
    ...
```
we'll give him prompt tokens count in any case

Reply:
```python
if not found_usage and (not error or completions):
    # append usage
```

is equivalent to:

```python
if not found_usage:
    if error and not completions:
        pass
    else:
        # append usage
```

Meaning that the **estimated usage** which hasn't been already reported in the stream isn't reported if:
(1) the server has returned an error AND,
(2) it didn't generate any completion content.

This may happen if the server returns 400, 422 or 500 right away.
It would be wrong to report `usage` _(and `usage.prompt_tokens` in particular)_, because the usage represents actual **payable** tokens. The combination of **(error & no completion generated)** is very unlikely _(though not impossible)_ to incur any charges. So we do not report them.

Compare it with the case of **(error & completion generation)**. It wouldn't be possible for the model to generate something without consuming `prompt_tokens` first AND we already have some generated completion, so we report it as `completion_tokens`.